File: semester-3/paradigms/labs/lab3.ml
(*Task 1*)
let iterate init pred func =
  let rec iter acc =
    if (pred acc) then iter (func acc)
    else acc
  in iter init;;

(* Task 2 a*)
let init func n =
  if n <= 0 then []
  else
    let seq = iterate [(n - 1)] (fun (hd::_) -> hd > 0) (fun acc -> ((List.hd acc) - 1)::acc)
    in List.map func seq;;

(*Task 2 b *)
let integral f a b n =
  let dx = (b -. a) /. (float_of_int n) in
  let func_vals = init (fun i -> f (a +. dx *. (float_of_int i))) (n + 1) in
  let rec calc_areas acc = function
    | [] | [_] -> acc
    | hd1::hd2::tail ->
      let area = ((hd1 +. hd2) /. 2.) *. dx
      in calc_areas (area::acc) (hd2::tail) in
  let sum = List.fold_left (+.) 0. in
  let areas = calc_areas [] func_vals in
  sum areas;;

"Task 1 (iterate)";;
iterate 0 (fun n -> n < 10) (fun n -> n + 1);;
iterate 0 (fun n -> n > 10) (fun n -> n + 1);;

"Task 2 A (init)";;
init (fun i -> i + 10) 10;;
init (fun i -> i + 10) 0;;
init (fun i -> i + 10) (-10);;

"Task 2 B (integral)";;

integral (fun x -> x) 0. 1. 10;;
integral (fun x -> x) 1. 0. 10;;
integral (fun x -> x) 0. 0. 10;;

integral sin 0. 3.14 1000;;

File: semester-3/paradigms/labs/lab3.scala
// Task 1
def iterate[A](init: A, pred: (A) => Boolean, func: (A) => A): A = {
  def iter(acc: A): A =
    if (pred(acc)) iter(func(acc))
    else acc

  iter(init)
}

// Task 2 A
def init[A](func: (Int) => A, n: Int): List[A] = {
  if (n < 0) Nil
  else {
    val seq = iterate(List(n-1), (acc: List[Int]) => acc.head > 0, (acc: List[Int]) => (acc.head - 1)::acc)
    seq.map(func)
  }
}


// Task 2 B
def integral(func: (Double) => Double, a: Double, b: Double, n: Int): Double = {
  val dx = (b - a) / n
  val func_vals = init(i => func(a + dx * i), n + 1)
  def calc_areas(acc: List[Double], vals: List[Double]): List[Double] =
    vals match {
      case Nil | List(_) => acc
      case hd1::hd2::tail => {
        val area = ((hd1 + hd2) / 2) * dx
        calc_areas(area::acc, hd2::tail)
      }
    }

  calc_areas(Nil, func_vals).sum
}

println("Task #1 (iterate)")
iterate(0, (n: Int) => n < 10, (n: Int) => n + 1)

println("Task #2 A (init)")
init(i => i + 1, 10)
init(i => i + 1, 0)
init(i => i + 1, -10)

println("Task #2 B (integral)")
integral(x => x, 0, 1, 10)
integral(x => x, 1, 0, 10)
integral(x => x, 0, 0, 10)

integral(Math.sin, 0, Math.PI, 1000)
